ReactSkia/core_modules/RSkSpatialNavigator.cpp

```diff
@@ -0,0 +1,380 @@
+* found in the LICENSE file.
+*/
+#include <string>
+#include <algorithm>
+
+#include "RSkSpatialNavigator.h"
+#include "ReactSkia/components/RSkComponent.h"
+#include "ReactSkia/sdk/RNSKeyCodeMapping.h"
+
+namespace SpatialNavigator {
+
+RSkSpatialNavigator* RSkSpatialNavigator::sharedSpatialNavigator_{nullptr};
+std::mutex RSkSpatialNavigator::mutex_;
+
+RSkSpatialNavigator::RSkSpatialNavigator() {
+    std::function<void(rnsKey, rnsKeyAction)> handler = std::bind(&RSkSpatialNavigator::handleKeyEvent, this,
+                                                                std::placeholders::_1, // rnsKey
+                                                                std::placeholders::_2);  // rnsKeyAction
+    eventId_ = NotificationCenter::defaultCenter().addListener("onHWKeyEvent", handler);
+}
+
+RSkSpatialNavigator::~RSkSpatialNavigator() {
+    if(this == sharedSpatialNavigator_)
+        sharedSpatialNavigator_ = nullptr;
+}
+
+RSkSpatialNavigator* RSkSpatialNavigator::sharedSpatialNavigator() {
+    std::lock_guard<std::mutex> lock(mutex_);
+    if (sharedSpatialNavigator_ == nullptr) {
+        sharedSpatialNavigator_ = new RSkSpatialNavigator();
+    }
+    return sharedSpatialNavigator_;
+}
+
+void RSkSpatialNavigator::sendNotificationWithEventType(std::string eventType, int tag, rnsKeyAction keyAction) {
+    if(eventType.c_str() == nullptr)
+        return;
+    RNS_LOG_DEBUG("Send : " << eventType  << " To ComponentTag : " << tag );
+    NotificationCenter::defaultCenter().emit("RCTTVNavigationEventNotification",
+                                                folly::dynamic(folly::dynamic::object("eventType", eventType.c_str())
+                                                                              ("eventKeyAction", (int)keyAction)
+                                                                              ("tag", tag)
+                                                                              ("target", tag)
+                                                                              ));
+}
+
+/*
+ * Add a new candidate to the Navigatable if it meets the following criteria:
+ * isTVSelectable, accessible are true
+*/
+void RSkSpatialNavigator::addToNavList(std::shared_ptr<RSkComponent> newCandidate) {
+    Component newCandData = newCandidate.get()->getComponentData();
+#if 0
+    if (candidate.props->accessible == true || candidate.isTVSelectable == true ) {
+#else
+    if(1) {
+#endif
+        navComponentList_.push_back(newCandidate.get());
+        // TODO If this new component has hasTVPreferredFocus then send Blur to current focused element and send focus to this new candidate
+        if(1) { //newCandidate.get()->props.hasTVPreferredFocus == true) {
+            if(currentFocus_)
+                sendNotificationWithEventType("blur", currentFocus_->getComponentData().tag);
+
+            sendNotificationWithEventType("Focus", newCandData.tag);
```

Comment:
the event name to be in camelcase "focus"?

Reply:
Done